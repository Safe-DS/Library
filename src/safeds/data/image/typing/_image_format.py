from enum import Enum


class ImageFormat(Enum):
    """Images formats supported by us."""

    JPEG = "jpeg"
    PNG = "png"
