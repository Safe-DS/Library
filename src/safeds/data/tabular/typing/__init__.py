"""Types used to define the schema of a tabular dataset."""

from ._column_type import Anything, Boolean, ColumnType, Integer, RealNumber, String
from ._imputer_strategy import ImputerStrategy
from ._schema import Schema

__all__ = [
    "Anything",
    "Boolean",
    "ColumnType",
    "ImputerStrategy",
    "Integer",
    "RealNumber",
    "Schema",
    "String",
]
