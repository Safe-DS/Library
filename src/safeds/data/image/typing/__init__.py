"""Types used to distinguish different image formats."""

from ._image_format import ImageFormat

__all__ = [
    "ImageFormat",
]
