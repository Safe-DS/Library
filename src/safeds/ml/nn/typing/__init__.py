"""Types used to define neural networks and related attributes."""

from typing import TYPE_CHECKING

import apipkg

if TYPE_CHECKING:
    from ._model_image_size import ConstantImageSize, ModelImageSize, VariableImageSize

apipkg.initpkg(
    __name__,
    {
        "ConstantImageSize": "._model_image_size:ConstantImageSize",
        "ModelImageSize": "._model_image_size:ModelImageSize",
        "VariableImageSize": "._model_image_size:VariableImageSize",
    },
)

__all__ = [
    "ConstantImageSize",
    "ModelImageSize",
    "VariableImageSize",
]
