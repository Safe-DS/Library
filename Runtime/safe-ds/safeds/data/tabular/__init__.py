from ._column import Column, ColumnStatistics
from ._row import Row
from ._table import Table
