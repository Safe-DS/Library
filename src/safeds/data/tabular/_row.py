import typing
from typing import Any

import pandas as pd
from IPython.core.display_functions import DisplayHandle, display
from safeds.data.tabular.typing import ColumnType, TableSchema
from safeds.exceptions import UnknownColumnNameError


class Row:
    def __init__(self, data: typing.Iterable, schema: TableSchema):
        self._data: pd.Series = data if isinstance(data, pd.Series) else pd.Series(data)
        self.schema: TableSchema = schema
        self._data = self._data.reset_index(drop=True)

    def __getitem__(self, column_name: str) -> Any:
        return self.get_value(column_name)

    def __iter__(self) -> typing.Iterator[Any]:
        return iter(self.get_column_names())

    def __len__(self) -> int:
        return len(self._data)

    def get_value(self, column_name: str) -> Any:
        """
        Return the value of a specified column.

        Parameters
        ----------
        column_name : str
            The column name.

        Returns
        -------
        value :
            The value of the column.
        """
        if not self.schema.has_column(column_name):
            raise UnknownColumnNameError([column_name])
        return self._data[self.schema._get_column_index_by_name(column_name)]

    def count(self) -> int:
        """
        Return the number of columns in this row.

        Returns
        -------
        count : int
            The number of columns.
        """
        return len(self._data)

    def has_column(self, column_name: str) -> bool:
        """
        Return whether the row contains a given column.
        Alias for self.schema.hasColumn(column_name: str) -> bool.

        Parameters
        ----------
        column_name : str
            The name of the column.

        Returns
        -------
        contains : bool
            True, if row contains the column.
        """
        return self.schema.has_column(column_name)

    def get_column_names(self) -> list[str]:
        """
        Return a list of all column names saved in this schema.
        Alias for self.schema.get_column_names() -> list[str].

        Returns
        -------
        column_names : list[str]
            The column names.
        """
        return self.schema.get_column_names()

    def get_type_of_column(self, column_name: str) -> ColumnType:
        """
        Return the type of a specified column.
        Alias for self.schema.get_type_of_column(column_name: str) -> ColumnType.

        Parameters
        ----------
        column_name : str
            The name of the column.

        Returns
        -------
        type : ColumnType
            The type of the column.

        Raises
        ------
        ColumnNameError
            If the specified target column name does not exist.
        """
        return self.schema.get_type_of_column(column_name)

    def __eq__(self, other: typing.Any) -> bool:
        if not isinstance(other, Row):
            return NotImplemented
        if self is other:
            return True
        return self._data.equals(other._data) and self.schema == other.schema

    def __hash__(self) -> int:
        return hash(self._data)

    def __str__(self) -> str:
        tmp = self._data.to_frame().T
        tmp.columns = self.get_column_names()
        return tmp.__str__()

    def __repr__(self) -> str:
        tmp = self._data.to_frame().T
        tmp.columns = self.get_column_names()
        return tmp.__repr__()

    def _ipython_display_(self) -> DisplayHandle:
        """
        Return a display object for the column to be used in Jupyter Notebooks.

        Returns
        -------
        output : DisplayHandle
            Output object.
        """
        tmp = self._data.to_frame().T
        tmp.columns = self.get_column_names()

        with pd.option_context(
            "display.max_rows", tmp.shape[0], "display.max_columns", tmp.shape[1]
        ):
            return display(tmp)
