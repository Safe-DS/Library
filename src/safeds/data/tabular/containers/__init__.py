from ._column import Column
from ._row import Row
from ._table import Table
from ._tagged_table import TaggedTable
