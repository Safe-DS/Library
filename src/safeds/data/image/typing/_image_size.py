from __future__ import annotations

import sys
from typing import TYPE_CHECKING

from safeds._utils import _structural_hash
from safeds.exceptions import ClosedBound, OutOfBoundsError

if TYPE_CHECKING:
    from safeds.data.image.containers import Image


class ImageSize:
    """
    A container for image size data.

    Parameters
    ----------
    width:
        the width of the image
    height:
        the height of the image
    channel:
        the channel of the image

    Raises
    ------
    OutOfBoundsError:
        if width or height are below 1
    ValueError
        if an invalid channel is given
    """

    def __init__(self, width: int, height: int, channel: int, *, _ignore_invalid_channel: bool = False) -> None:
        if width < 1 or height < 1:
            raise OutOfBoundsError(min(width, height), lower_bound=ClosedBound(1))
        elif not _ignore_invalid_channel and channel not in (1, 3, 4):
            raise ValueError(f"Channel {channel} is not a valid channel option. Use either 1, 3 or 4")
        elif channel < 1:
            raise OutOfBoundsError(channel, name="channel", lower_bound=ClosedBound(1))
        self._width = width
        self._height = height
        self._channel = channel

    @staticmethod
    def from_image(image: Image) -> ImageSize:
        """
        Create a `ImageSize` of a given image.

        Parameters
        ----------
        image:
            the given image for the `ImageSize`

        Returns
        -------
        image_size:
            the calculated `ImageSize`
        """
        return ImageSize(image.width, image.height, image.channel)

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, ImageSize):
            return NotImplemented
        return (self is other) or (
            self._width == other._width and self._height == other._height and self._channel == other._channel
        )

    def __hash__(self) -> int:
        return _structural_hash(self._width, self._height, self._channel)

    def __sizeof__(self) -> int:
        return sys.getsizeof(self._width) + sys.getsizeof(self._height) + sys.getsizeof(self._channel)

    def __str__(self) -> str:
        return f"{self._width}x{self._height}x{self._channel} (WxHxC)"

    @property
    def width(self) -> int:
        """
        Get the width of this `ImageSize` in pixels.

        Returns
        -------
        width:
            The width of this `ImageSize`.
        """
        return self._width

    @property
    def height(self) -> int:
        """
        Get the height of this `ImageSize` in pixels.

        Returns
        -------
        height:
            The height of this `ImageSize`.
        """
        return self._height

    @property
    def channel(self) -> int:
        """
        Get the channel of this `ImageSize` in pixels.

        Returns
        -------
        channel:
            The channel of this `ImageSize`.
        """
        return self._channel
