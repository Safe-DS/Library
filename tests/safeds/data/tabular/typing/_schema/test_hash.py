from collections.abc import Callable

import pytest
from syrupy import SnapshotAssertion

from safeds.data.tabular.typing import ColumnType, Schema


@pytest.mark.parametrize(
    "schema_factory",
    [
        lambda: Schema({}),
        lambda: Schema({"col1": ColumnType.null()}),
        lambda: Schema({"col1": ColumnType.null(), "col2": ColumnType.null()}),
    ],
    ids=[
        "empty",
        "one column",
        "two columns",
    ],
)
class TestContract:
    def test_should_return_same_hash_for_equal_objects(self, schema_factory: Callable[[], Schema]) -> None:
        schema_1 = schema_factory()
        schema_2 = schema_factory()
        assert hash(schema_1) == hash(schema_2)

    def test_should_return_same_hash_in_different_processes(
        self,
        schema_factory: Callable[[], Schema],
        snapshot: SnapshotAssertion,
    ) -> None:
        schema_ = schema_factory()
        assert hash(schema_) == snapshot


@pytest.mark.parametrize(
    ("schema_1", "schema_2"),
    [
        # not equal (too few columns)
        (
            Schema({"col1": ColumnType.null()}),
            Schema({}),
        ),
        # not equal (too many columns)
        (
            Schema({}),
            Schema({"col1": ColumnType.null()}),
        ),
        # not equal (different column order)
        (
            Schema({"col1": ColumnType.null(), "col2": ColumnType.int8()}),
            Schema({"col2": ColumnType.int8(), "col1": ColumnType.null()}),
        ),
        # not equal (different column names)
        (
            Schema({"col1": ColumnType.null()}),
            Schema({"col2": ColumnType.null()}),
        ),
        # not equal (different types)
        (
            Schema({"col1": ColumnType.null()}),
            Schema({"col1": ColumnType.int8()}),
        ),
    ],
    ids=[
        "too few columns",
        "too many columns",
        "different column order",
        "different column names",
        "different types",
    ],
)
def test_should_be_good_hash(
    schema_1: Schema,
    schema_2: Schema,
) -> None:
    assert hash(schema_1) != hash(schema_2)
