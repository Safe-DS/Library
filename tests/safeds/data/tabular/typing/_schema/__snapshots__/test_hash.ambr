# serializer version: 1
# name: TestContract.test_should_return_same_hash_in_different_processes[empty]
  2230328334128810753
# ---
# name: TestContract.test_should_return_same_hash_in_different_processes[one column]
  2245614720678699598
# ---
# name: TestContract.test_should_return_same_hash_in_different_processes[two columns]
  1881809609867026590
# ---
